File: testing/PIMS.Tests.Automation/PageObjects/AcquisitionAgreements.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using PIMS.Tests.Automation.Classes;
using SeleniumExtras.WaitHelpers;

namespace PIMS.Tests.Automation.PageObjects
{
    public class AcquisitionAgreements : PageObjectBase
    {
        //Acquisition Files Agreements Initial Elements
        private By agreementsLinkTab = By.XPath("//a[contains(text(),'Agreements')]");

        private By agreementsSubtitle = By.XPath("//h2/div/div/div/div[contains(text(),'Agreements')]");
        private By agreementsCreateNewAgreementBttn = By.XPath("//h2/div/div/div/div[contains(text(),'Agreements')]/following-sibling::div/button");
        private By agreementsInitMessage = By.XPath("//p[contains(text(),'There are no agreements indicated in this acquisition file.')]");

        //Acquisition Files Agreement Create Form Elements
        private By AgreementsDetailsCreateSubtitle = By.XPath("//h2/div/div[contains(text(),'Agreement Details')]");
        private By agreementsStatusLabel = By.XPath("//label[contains(text(),'Agreement status')]");
        private By agreementsStatusInput = By.Id("input-agreementStatusTypeCode");
        private By agreementCancellationReasonInput = By.Id("input-cancellationNote");
        private By agreementsLegalSurveyPlanLabel = By.XPath("//label[contains(text(),'Legal survey plan')]");
        private By agreementsLegalSurveyPlanInput = By.Id("input-legalSurveyPlanNum");
        private By agreementsTypeLabel = By.XPath("//label[contains(text(),'Agreement type')]");
        private By agreementsTypeSelect = By.Id("input-agreementTypeCode");
        private By agreementsDateLabel = By.XPath("//label[contains(text(),'Agreement date')]");
        private By agreementsDateInput = By.Id("datepicker-agreementDate");
        private By agreementCommencementDateInput = By.Id("datepicker-commencementDate");
        private By agreementsCompletionDateLabel = By.XPath("//label[contains(text(),'Completion date')]");
        private By agreementsCompletionDateInput = By.Id("datepicker-completionDate");
        private By agreementsTerminationDateLabel = By.XPath("//label[contains(text(),'Termination date')]");
        private By agreementsTerminationDateInput = By.Id("datepicker-terminationDate");
        private By agreementsPossessionDateLabel = By.XPath("//label[contains(text(),'Possession date')]");
        private By agreementsPossessionDateInput = By.Id("datepicker-possessionDate");

        private By agreementFinancialSubtitle = By.XPath("//div[contains(text(),'Financial')]");
        private By agreementsPurchasePriceLabel = By.XPath("//label[contains(text(),'Purchase price')]");
        private By agreementsPurchasePriceInput = By.Id("input-purchasePrice");
        private By agreementsDepositNotLaterLabel = By.XPath("//label[contains(text(),'Deposit due no later than')]");
        private By agreementsDepositNotLaterInput = By.Id("input-noLaterThanDays");
        private By agreementsDepositAmountLabel = By.XPath("//label[contains(text(),'Deposit amount')]");
        private By agreementsDepositAmountInput = By.Id("input-depositAmount");

        private By agreementsTotalCount = By.XPath("//div[contains(text(),'Agreements')]/parent::div/parent::div/parent::div/parent::h2/following-sibling::div[@class='collapse show']/div");


        //Acquisition File Confirmation Modal Elements
        private By acquisitionFileConfirmationModal = By.CssSelector("div[class='modal-content']");

        private SharedModals sharedModals;


        public AcquisitionAgreements(IWebDriver webDriver) : base(webDriver)
        {
            sharedModals = new SharedModals(webDriver);
        }

        public void NavigateAgreementsTab()
        {
            WaitUntilClickable(agreementsLinkTab);
            webDriver.FindElement(agreementsLinkTab).Click();
        }

        public void EditAgreementButton(int index)
        {
            Wait();
            var elementNumber = index + 1;
            webDriver.FindElement(By.CssSelector("button[data-testid='agreements["+ elementNumber +"].edit-btn']")).Click();
        }

        public void CreateNewAgreementBttn()
        {
            Wait();

            WaitUntilClickable(agreementsCreateNewAgreementBttn);
            webDriver.FindElement(agreementsCreateNewAgreementBttn).Click();
        }

        public void SaveAcquisitionFileAgreement()
        {
            Wait();
            ButtonElement("Save");

            AssertTrueIsDisplayed(agreementsCreateNewAgreementBttn);
        }

        public void CancelAcquisitionFileAgreement()
        {
            Wait();
            ButtonElement("Cancel");
           
            sharedModals.CancelActionModal();

            AssertTrueIsDisplayed(agreementsCreateNewAgreementBttn);
        }

        public void CreateUpdateAgreement(AcquisitionAgreement agreement)
        {
            Wait();

            ChooseSpecificSelectOption(agreementsStatusInput, agreement.AgreementStatus);

            if (agreement.AgreementCancellationReason != "")
            {
                ClearInput(agreementCancellationReasonInput);
                webDriver.FindElement(agreementCancellationReasonInput).SendKeys(agreement.AgreementCancellationReason);
            }

            if (agreement.AgreementLegalSurveyPlan != "")
            {
                ClearInput(agreementsLegalSurveyPlanInput);
                webDriver.FindElement(agreementsLegalSurveyPlanInput).SendKeys(agreement.AgreementLegalSurveyPlan);
            }

            if(agreement.AgreementType != "")
                ChooseSpecificSelectOption(agreementsTypeSelect, agreement.AgreementType);

            if (agreement.AgreementDate != "")
            {
                ClearInput(agreementsDateInput);
                webDriver.FindElement(agreementsDateInput).SendKeys(agreement.AgreementDate);
                webDriver.FindElement(agreementsDateInput).SendKeys(Keys.Enter);
            }

            if (agreement.AgreementCommencementDate != "")
            {
                ClearInput(agreementCommencementDateInput);
                webDriver.FindElement(agreementCommencementDateInput).SendKeys(agreement.AgreementCommencementDate);
                webDriver.FindElement(agreementCommencementDateInput).SendKeys(Keys.Enter);
            }

            if (agreement.AgreementCompletionDate != "")
            {
                ClearInput(agreementsCompletionDateInput);
                webDriver.FindElement(agreementsCompletionDateInput).SendKeys(agreement.AgreementCompletionDate);
                webDriver.FindElement(agreementsCompletionDateInput).SendKeys(Keys.Enter);
            }

            if (agreement.AgreementTerminationDate != "")
            {
                ClearInput(agreementsTerminationDateInput);
                webDriver.FindElement(agreementsTerminationDateInput).SendKeys(agreement.AgreementTerminationDate);
                webDriver.FindElement(agreementsTerminationDateInput).SendKeys(Keys.Enter);
            }

            if (agreement.AgreementPossessionDate != "")
            {
                ClearInput(agreementsPossessionDateInput);
                webDriver.FindElement(agreementsPossessionDateInput).SendKeys(agreement.AgreementPossessionDate);
                webDriver.FindElement(agreementsPossessionDateInput).SendKeys(Keys.Enter);
            }

            if (agreement.AgreementPurchasePrice != "")
            {
                ClearInput(agreementsPurchasePriceInput);
                webDriver.FindElement(agreementsPurchasePriceInput).SendKeys(agreement.AgreementPurchasePrice);
            }

            if (agreement.AgreementDepositDue != "")
            {
                ClearInput(agreementsDepositNotLaterInput);
                webDriver.FindElement(agreementsDepositNotLaterInput).SendKeys(agreement.AgreementDepositDue);
            }

            if (agreement.AgreementDepositAmount != "")
            {
                ClearInput(agreementsDepositAmountInput);
                webDriver.FindElement(agreementsDepositAmountInput).SendKeys(agreement.AgreementDepositAmount);
            }
        }

        public void DeleteLastAgreement()
        {
            var lastAgreement = webDriver.FindElements(agreementsTotalCount).Count();

            webDriver.FindElement(By.CssSelector("button[data-testid='agreements["+ lastAgreement +"].delete-btn']")).Click();

            if (webDriver.FindElements(acquisitionFileConfirmationModal).Count() > 0)
            {
                Assert.Equal("Delete Agreement", sharedModals.ModalHeader());
                Assert.Equal("You have selected to delete this Agreement.\r\n Do you want to proceed?", sharedModals.ModalContent());
                sharedModals.ModalClickOKBttn();
            }

        }

        public int TotalAgreementsCount()
        {
            Wait();
            return webDriver.FindElements(agreementsTotalCount).Count();
        }

        public void VerifyInitAgreementTab()
        {
            AssertTrueIsDisplayed(agreementsSubtitle);
            AssertTrueIsDisplayed(agreementsCreateNewAgreementBttn);
            AssertTrueIsDisplayed(agreementsInitMessage);
        }

        public void VerifyCreateAgreementForm()
        {
            AssertTrueIsDisplayed(AgreementsDetailsCreateSubtitle);

            AssertTrueIsDisplayed(agreementsStatusLabel);
            AssertTrueIsDisplayed(agreementsStatusInput);

            AssertTrueIsDisplayed(agreementsLegalSurveyPlanLabel);
            AssertTrueIsDisplayed(agreementsLegalSurveyPlanInput);

            AssertTrueIsDisplayed(agreementsTypeLabel);
            AssertTrueIsDisplayed(agreementsTypeSelect);

            AssertTrueIsDisplayed(agreementsDateLabel);
            AssertTrueIsDisplayed(agreementsDateInput);

            AssertTrueIsDisplayed(agreementsCompletionDateLabel);
            AssertTrueIsDisplayed(agreementsCompletionDateInput);

            AssertTrueIsDisplayed(agreementsTerminationDateLabel);
            AssertTrueIsDisplayed(agreementsTerminationDateInput);

            AssertTrueIsDisplayed(agreementsPossessionDateLabel);
            AssertTrueIsDisplayed(agreementsPossessionDateInput);

            AssertTrueIsDisplayed(agreementFinancialSubtitle);

            AssertTrueIsDisplayed(agreementsPurchasePriceLabel);
            AssertTrueIsDisplayed(agreementsPurchasePriceInput);

            AssertTrueIsDisplayed(agreementsDepositNotLaterLabel);
            AssertTrueIsDisplayed(agreementsDepositNotLaterInput);

            AssertTrueIsDisplayed(agreementsDepositAmountLabel);
            AssertTrueIsDisplayed(agreementsDepositAmountInput);
        }

        public void VerifyViewAgreementForm(AcquisitionAgreement agreement, int index)
        {
            var agreementNbr = index + 1;

            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/preceding-sibling::div"), "Agreement " + agreementNbr);

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Agreement status')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Agreement status')]/parent::div/following-sibling::div"), agreement.AgreementStatus);

            if (agreement.AgreementCancellationReason != "")
            {
                AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Cancellation reason')]"));
                AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Cancellation reason')]/parent::div/following-sibling::div"), agreement.AgreementCancellationReason);
            }

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Legal survey plan')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Legal survey plan')]/parent::div/following-sibling::div"), agreement.AgreementLegalSurveyPlan);

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Agreement type')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Agreement type')]/parent::div/following-sibling::div"), agreement.AgreementType);

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Agreement date')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Agreement date')]/parent::div/following-sibling::div"), TransformDateFormat(agreement.AgreementDate));

            if (agreement.AgreementCommencementDate != "")
            {
                AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Commencement date')]"));
                AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Commencement date')]/parent::div/following-sibling::div"), TransformDateFormat(agreement.AgreementCommencementDate));
            }

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Completion date')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Completion date')]/parent::div/following-sibling::div"), TransformDateFormat(agreement.AgreementCompletionDate));

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Termination date')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Termination date')]/parent::div/following-sibling::div"), TransformDateFormat(agreement.AgreementTerminationDate));

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Possession date')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Possession date')]/parent::div/following-sibling::div"), TransformDateFormat(agreement.AgreementPossessionDate));

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div[contains(text(),'Financial')]"));

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Purchase price')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Purchase price')]/parent::div/following-sibling::div"), TransformCurrencyFormat(agreement.AgreementPurchasePrice));

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Deposit due no later than')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Deposit due no later than')]/parent::div/following-sibling::div"), agreement.AgreementDepositDue + " days");

            AssertTrueIsDisplayed(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Deposit amount')]"));
            AssertTrueContentEquals(By.XPath("//button[@data-testid='agreements["+ agreementNbr +"].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'Deposit amount')]/parent::div/following-sibling::div"), TransformCurrencyFormat(agreement.AgreementDepositAmount));
        }
    }
}

File: testing/playwright-testing/support/pages/AcquisitionAgreements.js
const { expect } = require("@playwright/test");

class AcquisitionAgreements {
  constructor(page) {
    this.page = page;
  }

  async navigateAgreementsTab() {
    const agreementsLinkTab = await this.page.locator(
      "//a[contains(text(),'Agreements')]"
    );
    await agreementsLinkTab.waitFor({ state: "visible" });
    await agreementsLinkTab.click();
  }

  async editAgreementButton(index) {
    const elementNumber = index + 1;
    await this.page
      .locator(`button[data-testid='agreements[${elementNumber}].edit-btn']`)
      .click();
  }

  async createNewAgreement() {
    const agreementsCreateNewAgreementBtn = await this.page.locator(
      "//h2/div/div/div/div[contains(text(),'Agreements')]/following-sibling::div/button"
    );
    await agreementsCreateNewAgreementBtn.waitFor({ state: "visible" });
    await agreementsCreateNewAgreementBtn.click();
  }

  async saveAcquisitionFileAgreement() {
    await this.page.locator('button:has-text("Save")').click();
    const agreementsCreateNewAgreementBtn = await this.page.locator(
      "//h2/div/div/div/div[contains(text(),'Agreements')]/following-sibling::div/button"
    );
    expect(agreementsCreateNewAgreementBtn).toBeVisible();
  }

  async cancelAcquisitionFileAgreement() {
    await this.page.locator('button:has-text("Cancel")').click();
    const acquisitionFileConfirmationModal = await this.page.locator(
      "div[class='modal-content']"
    );

    if (await acquisitionFileConfirmationModal.isVisible()) {
      await this.page.locator('button:has-text("OK")').click();
    }

    const agreementsCreateNewAgreementBtn = await this.page.locator(
      "//h2/div/div/div/div[contains(text(),'Agreements')]/following-sibling::div/button"
    );
    await expect(agreementsCreateNewAgreementBtn).toBeVisible();
  }

  async createUpdateAgreement(agreement) {
    const agreementsStatusInput = await this.page.locator(
      "#input-agreementStatusTypeCode"
    );
    await agreementsStatusInput.selectOption({
      label: agreement.AgreementStatus,
    });

    if (agreement.AgreementCancellationReason) {
      const agreementCancellationReasonInput = await this.page.locator(
        "#input-cancellationNote"
      );
      await agreementCancellationReasonInput.fill(
        agreement.AgreementCancellationReason
      );
    }
    if (agreement.AgreementLegalSurveyPlan) {
      const agreementsLegalSurveyPlanInput = await this.page.locator(
        "#input-legalSurveyPlanNum"
      );
      await agreementsLegalSurveyPlanInput.fill(
        agreement.AgreementLegalSurveyPlan
      );
    }
    if (agreement.AgreementType) {
      const agreementsTypeSelect = await this.page.locator(
        "#input-agreementTypeCode"
      );
      await agreementsTypeSelect.selectOption({
        label: agreement.AgreementType,
      });
    }
    if (agreement.AgreementDate) {
      const agreementsDateInput = this.page.locator(
        "#datepicker-agreementDate"
      );
      await agreementsDateInput.fill(agreement.AgreementDate);
      await agreementsDateInput.press("Enter");
    }
    if (agreement.AgreementCommencementDate) {
      const agreementCommencementDateInput = await this.page.locator(
        "#datepicker-commencementDate"
      );
      await agreementCommencementDateInput.fill(
        agreement.AgreementCommencementDate
      );
      await agreementCommencementDateInput.press("Enter");
    }
    if (agreement.AgreementCompletionDate) {
      const agreementsCompletionDateInput = await this.page.locator(
        "#datepicker-completionDate"
      );
      await agreementsCompletionDateInput.fill(
        agreement.AgreementCompletionDate
      );
      await agreementsCompletionDateInput.press("Enter");
    }
    if (agreement.AgreementTerminationDate) {
      const agreementsTerminationDateInput = await this.page.locator(
        "#datepicker-terminationDate"
      );
      await agreementsTerminationDateInput.fill(
        agreement.AgreementTerminationDate
      );
      await agreementsTerminationDateInput.press("Enter");
    }
    if (agreement.AgreementPossessionDate) {
      const agreementsPossessionDateInput = await this.page.locator(
        "#datepicker-possessionDate"
      );
      await agreementsPossessionDateInput.fill(
        agreement.AgreementPossessionDate
      );
      await agreementsPossessionDateInput.press("Enter");
    }
    if (agreement.AgreementPurchasePrice) {
      const agreementsPurchasePriceInput = await this.page.locator(
        "#input-purchasePrice"
      );
      await agreementsPurchasePriceInput.fill(agreement.AgreementPurchasePrice);
    }
    if (agreement.AgreementDepositDue) {
      const agreementsDepositNotLaterInput = await this.page.locator(
        "#input-noLaterThanDays"
      );
      await agreementsDepositNotLaterInput.fill(agreement.AgreementDepositDue);
    }
    if (agreement.AgreementDepositAmount) {
      const agreementsDepositAmountInput = await this.page.locator(
        "#input-depositAmount"
      );
      await agreementsDepositAmountInput.fill(agreement.AgreementDepositAmount);
    }
  }

  async deleteLastAgreement() {
    const agreementsTotalCount = await this.page.locator(
      "//div[contains(text(),'Agreements')]/parent::div/parent::div/parent::div/parent::h2/following-sibling::div[@class='collapse show']/div"
    );
    const lastAgreement = await agreementsTotalCount.count();

    await this.page
      .locator(`button[data-testid='agreements[${lastAgreement}].delete-btn']`)
      .click();

    const acquisitionFileConfirmationModal = await this.page.locator(
      "div[class='modal-content']"
    );
    if (await acquisitionFileConfirmationModal.isVisible()) {
      await expect(this.page.locator("text=Delete Agreement")).toBeVisible();
      await expect(
        this.page.locator("text=You have selected to delete this Agreement.")
      ).toBeVisible();
      await this.page.click('button:has-text("OK")');
    }
  }

  async totalAgreementsCount() {
    const agreementsTotalCount = await this.page.locator(
      "//div[contains(text(),'Agreements')]/parent::div/parent::div/parent::div/parent::h2/following-sibling::div[@class='collapse show']/div"
    );
    return await agreementsTotalCount.count();
  }

  async verifyInitAgreementTab() {
    const agreementsSubtitle = await this.page.locator(
      "//h2/div/div/div/div[contains(text(),'Agreements')]"
    );
    await expect(agreementsSubtitle).toBeVisible();

    const agreementsCreateNewAgreementBtn = await this.page.locator(
      "//h2/div/div/div/div[contains(text(),'Agreements')]/following-sibling::div/button"
    );
    await expect(agreementsCreateNewAgreementBtn).toBeVisible();

    const agreementsInitMessage = await this.page.locator(
      "//p[contains(text(),'There are no agreements indicated in this acquisition file.')]"
    );
    await expect(agreementsInitMessage).toBeVisible();
  }

  async verifyCreateAgreementForm() {
    const AgreementsDetailsCreateSubtitle = await this.page.locator(
      "//h2/div/div[contains(text(),'Agreement Details')]"
    );
    await expect(AgreementsDetailsCreateSubtitle).toBeVisible();

    const agreementsStatusLabel = this.page.locator(
      "//label[contains(text(),'Agreement status')]"
    );
    await expect(agreementsStatusLabel).toBeVisible();

    const agreementsStatusInput = await this.page.locator(
      "#input-agreementStatusTypeCode"
    );
    await expect(agreementsStatusInput).toBeVisible();

    const agreementsLegalSurveyPlanLabel = await this.page.locator(
      "//label[contains(text(),'Legal survey plan')]"
    );
    await expect(agreementsLegalSurveyPlanLabel).toBeVisible();

    const agreementsLegalSurveyPlanInput = await this.page.locator(
      "#input-legalSurveyPlanNum"
    );
    await expect(agreementsLegalSurveyPlanInput).toBeVisible();

    const agreementsTypeLabel = await this.page.locator(
      "//label[contains(text(),'Agreement type')]"
    );
    await expect(agreementsTypeLabel).toBeVisible();

    const agreementsTypeSelect = await this.page.locator(
      "#input-agreementTypeCode"
    );
    await expect(agreementsTypeSelect).toBeVisible();

    const agreementsDateLabel = await this.page.locator(
      "//label[contains(text(),'Agreement date')]"
    );
    await expect(agreementsDateLabel).toBeVisible();

    const agreementsDateInput = await this.page.locator(
      "#datepicker-agreementDate"
    );
    await expect(agreementsDateInput).toBeVisible();

    const agreementsCompletionDateLabel = await this.page.locator(
      "//label[contains(text(),'Completion date')]"
    );
    await expect(agreementsCompletionDateLabel).toBeVisible();

    const agreementsCompletionDateInput = await this.page.locator(
      "#datepicker-completionDate"
    );
    await expect(agreementsCompletionDateInput).toBeVisible();

    const agreementsTerminationDateLabel = await this.page.locator(
      "//label[contains(text(),'Termination date')]"
    );
    await expect(agreementsTerminationDateLabel).toBeVisible();

    const agreementsTerminationDateInput = await this.page.locator(
      "#datepicker-terminationDate"
    );
    await expect(agreementsTerminationDateInput).toBeVisible();

    const agreementsPossessionDateLabel = await this.page.locator(
      "//label[contains(text(),'Possession date')]"
    );
    await expect(agreementsPossessionDateLabel).toBeVisible();

    const agreementsPossessionDateInput = await this.page.locator(
      "#datepicker-possessionDate"
    );
    await expect(agreementsPossessionDateInput).toBeVisible();

    const agreementFinancialSubtitle = await this.page.locator(
      "//div[contains(text(),'Financial')]"
    );
    await expect(agreementFinancialSubtitle).toBeVisible();

    const agreementsPurchasePriceLabel = await this.page.locator(
      "//label[contains(text(),'Purchase price')]"
    );
    await expect(agreementsPurchasePriceLabel).toBeVisible();

    const agreementsPurchasePriceInput = await this.page.locator(
      "#input-purchasePrice"
    );
    await expect(agreementsPurchasePriceInput).toBeVisible();

    const agreementsDepositNotLaterLabel = await this.page.locator(
      "//label[contains(text(),'Deposit due no later than')]"
    );
    await expect(agreementsDepositNotLaterLabel).toBeVisible();

    const agreementsDepositNotLaterInput = await this.page.locator(
      "#input-noLaterThanDays"
    );
    await expect(agreementsDepositNotLaterInput).toBeVisible();

    const agreementsDepositAmountLabel = await this.page.locator(
      "//label[contains(text(),'Deposit amount')]"
    );
    await expect(agreementsDepositAmountLabel).toBeVisible();

    const agreementsDepositAmountInput = await this.page.locator(
      "#input-depositAmount"
    );
    await expect(agreementsDepositAmountInput).toBeVisible();
  }

  async verifyViewAgreementForm(agreement, index) {
    const agreementNbr = index + 1;
    // Helper to get dynamic XPath with agreementNbr
    const x = (label) =>
      `//button[@data-testid='agreements[${agreementNbr}].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'${label}')]/parent::div/following-sibling::div`;
    const xLabel = (label) =>
      `//button[@data-testid='agreements[${agreementNbr}].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div/div/label[contains(text(),'${label}')]`;

    // Agreement header
    await expect(
      this.page.locator(
        `//button[@data-testid='agreements[${agreementNbr}].edit-btn']/parent::div/parent::div/preceding-sibling::div`
      )
    ).toHaveText(`Agreement ${agreementNbr}`);

    // Agreement status
    await expect(this.page.locator(xLabel("Agreement status"))).toBeVisible();
    await expect(this.page.locator(x("Agreement status"))).toHaveText(
      agreement.AgreementStatus
    );

    // Cancellation reason (optional)
    if (agreement.AgreementCancellationReason) {
      await expect(
        this.page.locator(xLabel("Cancellation reason"))
      ).toBeVisible();
      await expect(this.page.locator(x("Cancellation reason"))).toHaveText(
        agreement.AgreementCancellationReason
      );
    }

    // Legal survey plan
    await expect(this.page.locator(xLabel("Legal survey plan"))).toBeVisible();
    await expect(this.page.locator(x("Legal survey plan"))).toHaveText(
      agreement.AgreementLegalSurveyPlan
    );

    // Agreement type
    await expect(this.page.locator(xLabel("Agreement type"))).toBeVisible();
    await expect(this.page.locator(x("Agreement type"))).toHaveText(
      agreement.AgreementType
    );

    // Agreement date
    await expect(this.page.locator(xLabel("Agreement date"))).toBeVisible();
    await expect(this.page.locator(x("Agreement date"))).toHaveText(
      this.transformDateFormat(agreement.AgreementDate)
    );

    // Commencement date (optional)
    if (agreement.AgreementCommencementDate) {
      await expect(
        this.page.locator(xLabel("Commencement date"))
      ).toBeVisible();
      await expect(this.page.locator(x("Commencement date"))).toHaveText(
        this.transformDateFormat(agreement.AgreementCommencementDate)
      );
    }

    // Completion date
    await expect(this.page.locator(xLabel("Completion date"))).toBeVisible();
    await expect(this.page.locator(x("Completion date"))).toHaveText(
      this.transformDateFormat(agreement.AgreementCompletionDate)
    );

    // Termination date
    await expect(this.page.locator(xLabel("Termination date"))).toBeVisible();
    await expect(this.page.locator(x("Termination date"))).toHaveText(
      this.transformDateFormat(agreement.AgreementTerminationDate)
    );

    // Possession date
    await expect(this.page.locator(xLabel("Possession date"))).toBeVisible();
    await expect(this.page.locator(x("Possession date"))).toHaveText(
      this.transformDateFormat(agreement.AgreementPossessionDate)
    );

    // Financial subtitle
    await expect(
      this.page.locator(
        `//button[@data-testid='agreements[${agreementNbr}].edit-btn']/parent::div/parent::div/parent::div/parent::div/parent::div/parent::h2/following-sibling::div/div[contains(text(),'Financial')]`
      )
    ).toBeVisible();

    // Purchase price
    await expect(this.page.locator(xLabel("Purchase price"))).toBeVisible();
    await expect(this.page.locator(x("Purchase price"))).toHaveText(
      this.transformCurrencyFormat(agreement.AgreementPurchasePrice)
    );

    // Deposit due no later than
    await expect(
      this.page.locator(xLabel("Deposit due no later than"))
    ).toBeVisible();
    await expect(this.page.locator(x("Deposit due no later than"))).toHaveText(
      `${agreement.AgreementDepositDue} days`
    );

    // Deposit amount
    await expect(this.page.locator(xLabel("Deposit amount"))).toBeVisible();
    await expect(this.page.locator(x("Deposit amount"))).toHaveText(
      this.transformCurrencyFormat(agreement.AgreementDepositAmount)
    );
  }
}

module.exports = { AcquisitionAgreements };
